odoo/tools/safe_eval.py

```diff
@@ -238,131 +605,144 @@ def const_eval(expr):
-    :throws SyntaxError: If the expression provided is not valid Python
-    :throws NameError: If the expression provided accesses forbidden names
-    :throws ValueError: If the expression provided uses forbidden bytecode
+    Returns:
+        type: The type that is allowed
     """
-    if type(expr) is CodeType:
-        raise TypeError("safe_eval does not allow direct evaluation of code objects.")
-
-    # prevent altering the globals/locals from within the sandbox
-    # by taking a copy.
-    if not nocopy:
-        # isinstance() does not work below, we want *exactly* the dict class
-        if (globals_dict is not None and type(globals_dict) is not dict) \
-                or (locals_dict is not None and type(locals_dict) is not dict):
-            _logger.warning(
-                "Looks like you are trying to pass a dynamic environment, "
-                "you should probably pass nocopy=True to safe_eval().")
-        if globals_dict is not None:
-            globals_dict = dict(globals_dict)
-        if locals_dict is not None:
-            locals_dict = dict(locals_dict)
-
-    check_values(globals_dict)
-    check_values(locals_dict)
+    global _wrapped_types  # noqa: PLW0603 (Reason: we need to modify the global variable, check docstring) # pylint: disable=global-statement
 
+    _wrapped_types = _wrapped_types + (t,)
+    return allow_instance(t)
+
+
+def test_python_expr(expr: str, mode: str = "eval") -> str | bool:
+    """Test a Python expression before evaluating it
+
+    Args:
+        expr (str): A Python expression
+        mode (str): The mode of the expression (eval, exec)
+
+    Returns:
+        str | bool: An error message if the expression is not safe, False otherwise
+    """
+    try:
+        wrapped_expr = ast.unparse(
+            CodeChecker(_AST_ALLOWED_NODES).visit(ast.parse(expr))
+        )
+        compile(wrapped_expr, "<sandbox>", mode)
+    except (TypeError, NameError, SyntaxError, ValueError) as err:
+        if len(err.args) >= 2 and len(err.args[1]) >= 4:
+            error = {
+                "message": err.args[0],
+                "filename": err.args[1][0],
+                "lineno": err.args[1][1],
+                "offset": err.args[1][2],
+                "error_line": err.args[1][3],
+            }
+
+            msg = f"{type(err).__name__} : {error['message']} at line {error['lineno']}\n{error['error_line']}"
+        else:
+            msg = f"{err}"
+        return msg
+    return False
+
+
+def safe_eval(
+    expr: str,
+    globals_dict: Optional[ContextType] = None,
+    locals_dict: Optional[ContextType] = None,
+    mode: Literal["eval", "exec"] = "eval",
+    filename: Optional[str] = None,
+    locals_builtins: bool = False,
+    sandbox_instances: OptionalTuple[type] = (),
+    sandbox_types: OptionalTuple[type] = (),
+    ast_subset: (
+        frozenset[type[ast.AST]] | tuple[type[ast.AST], ...]
+    ) = _AST_ALLOWED_NODES,  # pylint: disable=bad-whitespace
+):
```

Comment:
This function and derivatives are missing a docstring listing what is allowed and not allowed in a human-readable way